### PR TITLE
Add QOA (Quite OK Audio) as a WAV compression mode

### LINKED ISSUE #9133: Add support for QOA (Quite OK Audio) format
*Related to #1144, #4264, #7599 and godotengine/godot#80160*.

### Describe the project you are working on

Game with sound effects.

### Describe the problem or limitation you are having in your project

The options we currently have on the engine for sound effects are:
- WAV 16-bit: loses on size;
- Vorbis and MP3: loses on decoding speed, making it not viable for lots of sound effects playing at the same time;
- WAV IMA-ADPCM: loses *drastically* on quality;

ADPCM-XQ somewhat makes IMA-ADPCM better, but the resulting audio will still have noticeable artifacts;

About the same "problem or limitation" as #7599 (ADPCM-XQ proposal).

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add support for the [Quite OK Audio Format (QOA)](https://qoaformat.org/).

When it comes to characteristics above:
- Size: smaller than IMA-ADPCM (3.2 bits vs 4 bits);
- Decoding speed: about the same as IMA-ADPCM;
- Quality: while it somewhat loses to MP3 and Vorbis (mostly due to a noise that's also somewhat present in ADPCM-XQ, it's still much better than IMA-ADPCM;

This makes it an ideal lossy compressed format for sound effects.

#### Approach 1: QOA file importing

Users would be able to import QOA files like they would do with WAV, MP3 and Ogg.
One problem with this approach however is that no common software nowadays have support for exporting files as QOA, and there's little interest in implementing it. At the time of this writing, FFmpeg has only implemented a decoder.

#### Approach 2: Import WAVs as QOA

Enable a Quite OK Audio importer on the Import As dropdown for WAVs:

image

It's an implementation of a built-in converter. This way, an user could work with WAVs within the game files, but internally it will be imported as QOA.

Add QOA as a WAV compression mode:

image

*There is a bit of resistance when it comes to implementing encoders inside Godot, but for QOA I believe it makes sense since QOA conversion isn't supported by major file converters and there's little interest on it*

Approaches 1 and 2 have been used in [a custom working module I made](https://github.com/DeeJayLSP/godot-qoa-module) (the screenshot above is from it).
It adds around 8kB into release template file size, but this gets mitigated as long as at least 0.06-0.13s of WAV audio gets QOA'd.

Approach 3, on the other hand, only had a binary size increase of 32 bytes.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Unlikely.

### Is there a reason why this should be core and not an add-on in the asset library?

In the case of Approach 3 and considering there's some support for it, I don't think an importer with negligible size penalty meant for compression should be an addon.

### PR DESCRIPTION
**Pull request by DeeJayLSP**
This is an alternative (a better one in my opinion) to #88646, with all caveats from it nullified. Closes godotengine/godot-proposals#9133 too.

Once again, QOA was developed to (not exactly just this, but is the best case) be a better alternative to ADPCM formats for use in games according to [the article announcing it](https://phoboslab.org/log/2023/02/qoa-time-domain-audio-compression).

Production release templates, for some reason, had a binary size **decrease** of 4064 bytes on Linux.

The patches in `qoa.h` suppresses a few warnings, allow editor to build (due to implementation being applied in both importer and stream) and reduce binary size penalty (it would be a bit over 4KiB otherwise).

---
This simply adds QOA as a compression mode within AudioStreamWAV:

image

Briefly, the differences between IMA-ADPCM and QOA within AudioStreamWAV should be:

*\+* IMA-ADPCM distorts lots of sound types, specially higher frequencies. The maximum QOA will do is add a barely audible white noise to higher frequencies.

*\+* IMA-ADPCM isn't resampled on playback, which means sounds different than the project's mix rate will get incredibly distorted. QOA doesn't use prediction when fetching decoded samples, so it can be resampled.

*\+* Since IMA-ADPCM decoding uses prediction, only Forward loop mode is available. While QOA does use prediction too, it does within frames that are decoded to a buffer and fetches samples from that, so all loop modes can work. Resampling had to be adapted to avoid some unnecessary decode callbacks.

*\-* QOA is slightly more complex than IMA-ADPCM, which should result in increased CPU usage. Despite this, it's still much faster than MP3 and Vorbis.

Supersedes #88646. Unlike it, QOA files can't be used (which shouldn't be a problem, as IMA-ADPCM WAVs could never be used either).

### CODE REVIEW

File: doc/classes/AudioStreamWAV.xml [hunk 1]
```diff
@@ -56,6 +56,9 @@
 		<constant name="FORMAT_IMA_ADPCM" value="2" enum="Format">
 			Audio is compressed using IMA ADPCM.
 		</constant>
+		<constant name="FORMAT_QOA" value="3" enum="Format">
+			Audio is compressed as QOA.
```
Comment (Calinou) [inline doc/classes/AudioStreamWAV.xml:60]:
```suggestion
			Audio is compressed as QOA ([url=https://qoaformat.org/]Quite OK Audio[/url]).
```

File: doc/classes/ResourceImporterWAV.xml [hunk 2]
```diff
@@ -13,7 +13,8 @@
 		<member name="compress/mode" type="int" setter="" getter="" default="0">
 			The compression mode to use on import.
 			[b]Disabled:[/b] Imports audio data without any compression. This results in the highest possible quality.
-			[b]RAM (Ima-ADPCM):[/b] Performs fast lossy compression on import. Low CPU cost, but quality is noticeably decreased compared to Ogg Vorbis or even MP3.
+			[b]RAM (IMA-ADPCM):[/b] Performs fast lossy compression on import. Low CPU cost, but quality is noticeably decreased compared to Ogg Vorbis or even MP3.
+			[b]QOA (Quite OK Audio):[/b] Performs lossy compression on import. CPU cost is slightly higher than IMA-ADPCM, but quality decrease is much lower.
```
Comment (Calinou) [inline doc/classes/ResourceImporterWAV.xml:17]:
```suggestion
			[b]QOA ([url=https://qoaformat.org/]Quite OK Audio[/url]):[/b] Performs lossy compression on import. CPU cost is slightly higher than IMA-ADPCM, but quality decrease is much lower.
```

**Review by Calinou — APPROVED** [review]
Tested locally, it works as expected. Same MRP as here: https://github.com/godotengine/godot/pull/88646#pullrequestreview-1917839320

File sizes of the imported WAV (converted to QOA on import) match the one from the previous PR.

**Comment by DeeJayLSP** [comment]
> I really like this approach! (can we call this non-destructive asset management?)
> 
> But at this point I wonder why we can't have the best of both worlds and also allow importing plain QOA files, basically merging this and the superseded PR. Sorry, was there a reason for that? I can't find it in the old thread. There seemed to be a pretty good consensus there.

Currently there is no software capable of converting audio files into QOA for distribution. Therefore few people would use it.

Also, according to the format's creator, QOA is meant to be embeddable.

**Review by deralmas — COMMENTED** [review]
Just took a look, and everything looks fine! I'm not an audio person so some thing flew over my head but they didn't raise any obvious red flags.

Small nitpick: the patch comments a bit of stuff. I suppose that those lines might as well be removed altogether.

File: thirdparty/misc/patches/qoa-min-fix.patch [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+ 			int slice_end = (sample_index + slice_len) * channels + c;			
+@@ -391,10 +391,10 @@ unsigned int qoa_encode_frame(const short *sample_data, qoa_desc *qoa, unsigned
+ 			16 scalefactors, encode all samples for the current slice and 
+ 			meassure the total squared error. */
+ 			qoa_uint64_t best_rank = -1;
+-			qoa_uint64_t best_error = -1;
+-			qoa_uint64_t best_slice;
+-			qoa_lms_t best_lms;
+-			int best_scalefactor;
++			//qoa_uint64_t best_error = -1;
```
[the remark this answers is not in the document]
Comment (DeeJayLSP) [reply thirdparty/misc/patches/qoa-min-fix.patch]: For some reason I thought the diff would be smaller if I commented instead of removing.

**Comment by DeeJayLSP** [comment]
The amount of workarounds I have to do for the sake of resampling

**Comment by DeeJayLSP** [comment]
I explained this a few times but I wanted to leave a definitive explanation for the workarounds.

QOA frames are composed of 5120 samples. The moment it begins playing, or when it goes from 5119 to 5120, the decoder is triggered, and a buffer large enough for 5120 samples gets replaced with the new frame data.

PCM8/16's resampling works by interpolating the current sample with the next. And this is a problem if we're trying to interpolate a backwards playback or a different sampling rate.

On a backwards playback, the following situation could occur:
```
// Previous sample: 5120
- Fetch sample 5119, interpolate with 5120
- Fetch sample 5118, interpolate with 5119
```
It crossed the 5120 interval 3 times, therefore **the decoder would be triggered 3 times for two samples in a row**.

The solution was simple: fetch samples backwards. Add + 1 to the current position, then reverse the from/to assignment.

I don't know if the other problem affected higher sample rates but surely did on lower. Due to the way it works, some samples end up repeating, with different fractions. If the repeated sample happens to be the last one in a QOA frame the same problem above would occur.

```
- Fetch sample 5119, interpolate with 5120
- Fetch sample 5119, interpolate with 5120 // Again but with a different fraction
```

The solution I came up with was to store the current sample, then return it back if the next request is for the same. I don't think this is a good solution but at least solves the problem.

File: scene/resources/audio_stream_wav.cpp [hunk 4]
```diff
@@ -175,32 +178,77 @@ void AudioStreamPlaybackWAV::do_resample(const Depth *p_src, AudioFrame *p_dst,
...
+						final_r = p_qoa->last_pos_final_r;
+					}
 
-			if constexpr (sizeof(Depth) == 1) { /* conditions will not exist anymore when compiled! */
-				final <<= 8;
+					next = p_qoa->last_pos_next;
+					if (is_stereo) {
+						next_r = p_qoa->last_pos_next_r;
+					}
+				}
```
Comment (DeeJayLSP) [inline scene/resources/audio_stream_wav.cpp:224]: Without resampling, this `is_qoa` block would be just this:
```cpp
			if (is_qoa) {
				uint32_t new_data_ofs = 8 + pos / QOA_FRAME_LEN * p_qoa->frame_len;
				int ch = p_qoa->desc->channels;

				if (p_qoa->data_ofs != new_data_ofs) {
					p_qoa->data_ofs = new_data_ofs;
					const uint8_t *src_ptr = (const uint8_t *)base->data;
					src_ptr += p_qoa->data_ofs + AudioStreamWAV::DATA_PAD;
					qoa_decode_frame(src_ptr, p_qoa->frame_len, p_qoa->desc, p_qoa->dec, &p_qoa->dec_len);
				}

				uint32_t dec_idx = (pos % QOA_FRAME_LEN) * ch;

				final = p_qoa->dec[dec_idx];
				if (is_stereo) {
					final_r = p_qoa->dec[dec_idx + 1];
				}
			}
```

Comment (DeeJayLSP) [reply scene/resources/audio_stream_wav.cpp:224]: Added some brief comments.

File: scene/resources/audio_stream_wav.cpp [hunk 5]
```diff
@@ -175,32 +178,77 @@ void AudioStreamPlaybackWAV::do_resample(const Depth *p_src, AudioFrame *p_dst,
 			}
 
 		} else {
-			final = p_src[pos];
-			if (is_stereo) {
-				final_r = p_src[pos + 1];
-			}
+			if (is_qoa) {
+				if (pos != p_qoa->last_pos) { // Prevents triple decoding on lower mix rates
```
Comment (AThousandShips) [inline scene/resources/audio_stream_wav.cpp:182]:
```suggestion
				if (pos != p_qoa->last_pos) { // Prevents triple decoding on lower mix rates.
```

File: scene/resources/audio_stream_wav.cpp [hunk 6]
```diff
@@ -175,32 +178,77 @@ void AudioStreamPlaybackWAV::do_resample(const Depth *p_src, AudioFrame *p_dst,
...
+					for (int i = 0; i < 2; i++) {
+						// Sign operations prevent triple decoding on backward loops, maxing prevents pop
```
Comment (AThousandShips) [inline scene/resources/audio_stream_wav.cpp:184]:
```suggestion
						// Sign operations prevent triple decoding on backward loops, maxing prevents pop.
```

**Review by AThousandShips — APPROVED** [review]

File: editor/import/resource_importer_wav.cpp [hunk 7]
```diff
@@ -28,6 +28,9 @@
 /* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
 /**************************************************************************/
 
+#define QOA_IMPLEMENTATION
+#define QOA_NO_STDIO
```
Comment (akien-mga) [inline editor/import/resource_importer_wav.cpp:32]: This seems weird. Single header library usually require such defines only once, before including the header.
The header isn't included explicitly here (nor in `audio_stream_wav.cpp`), but in `audio_stream_wav.h`, so I think these defines should be moved to `audio_stream_wav.h` only.

**Comment by DeeJayLSP** [comment]
Is it normal for binary sizes to *decrease* by over 4000 bytes after implementing a feature?

**Comment by DeeJayLSP** [comment]
For a while I'm unable to find ways to optimize/fix potential problems in this implementation, so this is the part I say I believe it's in the best state.

**Review by reduz — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**